File: other_scripts/minify.py
#!/usr/bin/env python
import os
import sys
import shutil
from datetime import date, timedelta
from subprocess import check_call, CalledProcessError
class JSHandler(object):
	def __init__(self, infile):
		self.infile = infile
		self.outfile = infile.replace(".js",".min.js")
		self.file_to_check = self.outfile
	def esvalidate(self):
		try:
			check_call("esvalidate %s" % self.infile, shell=True)
			print ("esvalidate succeeded")
			return True
		except CalledProcessError as e:
			print("esvalidate failed")
			print(e)
			return False
	def release(self):
		if not os.path.exists("releases"):
			os.makedirs("releases")
		dst = os.path.join("releases", date.today().strftime("%Y-%m-%d") + "-" + self.infile)
		print (dst)
		shutil.copyfile(self.infile, dst)
	def terser(self, outfile=None):
		if outfile is not None:
			self.outfile = outfile
		try:
			check_call("terser %s --compress sequences=true,conditionals=true,booleans=true --mangle --ecma 6 --output %s" % (self.infile, self.outfile), shell=True)
			print ("minification succeeded")
			return True
		except CalledProcessError as e:
			print("minification failed")
			print(e)
			return False
	def node_check(self,file_to_check=None):
		if file_to_check is not None:
			self.file_to_check = file_to_check
		try:
			check_call("node --check %s" % self.file_to_check, shell=True)
			print ("node check succeeded")
			return True
		except CalledProcessError as e:
			print("node check failed")
			print(e)
			return False
	def minify(self, outfile=None):
		if outfile is not None:
			self.outfile = outfile
			self.file_to_check = outfile
		if(self.esvalidate()):
			if(self.terser()):
				if (self.node_check()):
					print("========================================")
					print("%s successfully generated." % self.outfile)
					print("========================================")
					self.release()
if __name__ == '__main__':
	if len(sys.argv) == 2:
		JSHandler(sys.argv[1]).minify()
	elif len(sys.argv) == 3:
		JSHandler(sys.argv[1]).minify(sys.argv[2])
	else:
		print("Usage: minify.py infile [outfile]")

File: other_scripts/minify.mjs
#!/usr/bin/env zx

//import 'zx/globals';

// Function to display error messages and exit
function exitWithError(message) {
    console.error(message);
    process.exit(1);
  }
  
  // Get the current date
  let d = new Date().toISOString().split('T')[0];
  
  // Check for arguments
  if (process.argv.length < 4) {
    exitWithError("You must enter a JavaScript file name!");
  }
  
  let infile = process.argv[3];
  
  // Check if the input file exists and is a JavaScript file
  if (!fs.existsSync(infile)) {
    exitWithError(`${infile} does not exist!`);
  }
  
  if (!infile.endsWith('.js')) {
    exitWithError(`${infile} is not a JavaScript file!`);
  }
  
  let outfile;
  if (process.argv.length < 5) {
    outfile = infile.replace(/\.js$/, '.min.js');
  } else {
    outfile = process.argv[4];
    
    if (infile === outfile) {
      exitWithError("Please don't override your source code!");
    }
    
    if (!outfile.endsWith('.min.js')) {
      exitWithError(`${outfile} for security reasons, the output extension has to be .min.js!`);
    }
  }
  
  // Create the releases directory if it doesn't exist
  if (!fs.existsSync('releases')) {
    await $`mkdir releases`;
  }
  
  let fulldir = path.resolve(infile);
  
  
  
  try {
    await $`acorn --silent --ecma2018 ${infile}`;
    console.log(`acorn succeeded`);
  } catch {
    exitWithError(`acorn failed`);
  }
  
  try {
    await $`jshint --verbose ${infile}`;
    console.log(`jshint succeeded`);
  } catch {
    exitWithError(`jshint failed`);
  }
  
  try {
    await $`terser ${infile} --compress sequences=true,conditionals=true,booleans=true --mangle --ecma 6 --output ${outfile}`;
    console.log(`minification succeeded`);
  } catch {
    exitWithError(`minification failed`);
  }
  
  await $`sed -i 's/\\* @preserve/\\*/g' ${outfile}`;
  console.log("removing @preserve keyword from multiline comments");
  
  try {
    await $`node --check ${outfile}`;
    console.log(`node check succeeded`);
    await $`cp ${infile} releases/${d}-${path.basename(infile)}`;
  } catch {
    exitWithError(`node check failed`);
  }
  
  if (fs.existsSync(outfile)) {
    console.log("========================================");
    console.log(`${outfile} successfully generated.`);
    console.log("========================================");
    console.log("Tip: in order to preserve multiline comments, add @preserve to them.");
  }
  